pkgs/checks/lib/src/collection_equality.dart

```diff
@@ -0,0 +1,291 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:collection';
+
+import 'package:checks/context.dart';
+
+/// Returns a rejection if the elements of [actual] are unequal to the elements
+/// of [expected].
+///
+/// {@template deep_collection_equals}
+/// Elements, keys, or values, which are a collections are deeply compared for
+/// equality, and do not use the native identity based equality or custom
+/// equality operator overrides.
+/// Elements, keys, or values, which are a [Condition] instances are checked
+/// against actual values.
+/// All other value or key types use `operator ==`.
+///
+/// Does not use [Set.contains] or [Map.containsKey]. Custom collection behavior
+/// is ignored. For example, it is not possible to distinguish between a `Set`
+/// and a `Set.identity`.
+/// {@endtemplate}
+Rejection? deepCollectionEquals(Object actual, Object expected) {
```

Comment:
We should also probably support an unorderedEquals, at least for iterables? The logic could probably be shared with the set equality logic?

Reply:
I think unordered equality and deep equality should be separate conditions - I don't think `matcher` supports combining unordered and deep equality. I'll add a shallow `unorderedEquals` in another PR, and we can hold `unorderedDeepEquals` until we find a compelling use case.